### PR TITLE
Plugin-in transformation configuration and wider transformation usage

### LINKED ISSUE #2123: How should we configure transformations?
Currently the specification and configuration the transformation are done in the input/output itself. This leads to a lot of cross validation, when the options are in reality mutually exclusive. We should look into a better way of doing this.

### LINKED ISSUE #2764: On the usage of mimes for validation of data passing
**Introduction**
Currently, there is a `mime` type attached to input records of an ensemble (configured in `ensemble.yml`) and for input and output records of all steps (configured in `stages.yml`). This information is used to validate that the data being passed into the forward model (root stage) is consistent with the expected data of that stage. This is however not the intended usage of mimes and should be changed.

**The role of mimes**
A mime should be used to indicate serialisation and deserialisation format at hand when loading from disk or persisting to disk.

**Consequences**
Hence, it makes sense to specify a mime type when loading input records from disk. That is, when the `ert3 record load` functionality is being used or the `resources.<some-resource>` in `ensemble.yml`. This indicates the data format of the content that is to be loaded. However, when records are fetched from storage or `parameters.yml` via the `storage.<record-name>` or `stochastic.<parameter-name>` it makes no sense to dictate the mime type as the data has already, or will be created, under the hood without any interaction with disk.

Similarly, it makes sense that a unix step has to specify the mime type of all its input and output records as they are respectively being persisted to disk or loaded from disk. But, a function step is not doing this and should hence not have mime types configured for neither its input nor output records.

This also entails that it should be perfectly fine to have a different mime types when loading records prior to the ensemble evaluation and when persisting the same record in a unix step. That is, you could load a record persisted as json and write it as yaml (or csv whenever we support that).

Notice that it is the role of the transformation to convert records between its record representation and whatever the mime type indicates that it can be serialised to (or from).

Last, the validation of record types when being passed into the forward model (configured in `ensemble.yml`) or between two steps it should be the consistency of record type that is validated,  not the specified mime types.

### PR DESCRIPTION
**Pull request by jondequinor**
**Issue**
Resolves #2123
Resolves #2764

**Approach**
_See commit_

## Pre review checklist

- [x] Added appropriate labels

Adding labels helps the maintainers when writing release notes, see sections and the
corresponding labels here: https://github.com/equinor/ert/blob/main/.github/release.yml

### CODE REVIEW

File: ert/data/record/_record.py [hunk 1]
```diff
@@ -466,38 +471,31 @@ def collection_type(self) -> RecordCollectionType:
 
 
 async def load_collection_from_file(
-    file_path: pathlib.Path,
-    mime: str,
+    transformation: "RecordTransformation",
     length: int = 1,
-    is_directory: bool = False,
-    smry_keys: Optional[List[str]] = None,
+    root_path: pathlib.Path = pathlib.Path(),
```
Comment (jondequinor) [inline ert/data/record/_record.py:476]: Make optional, needs to be None as an object in initializer list is an anti-pattern.

File: ert/data/record/_transformation.py [hunk 2]
```diff
@@ -47,89 +90,141 @@ async def _make_tar(file_path: Path) -> bytes:
     return await get_event_loop().run_in_executor(executor, _sync_make_tar, file_path)
 
 
+class RecordTransformationAffinity(Flag):
+    OPAQUE = auto()
+    NON_OPAQUE = auto()
+    NONE = OPAQUE | NON_OPAQUE
```
Comment (jondequinor) [inline ert/data/record/_transformation.py:96]: might be removed

File: ert/data/record/_transformation.py [hunk 3]
```diff
@@ -47,89 +90,141 @@ async def _make_tar(file_path: Path) -> bytes:
...
         if not isinstance(record, (NumericalRecord, BlobRecord)):
             raise TypeError("Record type must be a NumericalRecord or BlobRecord")
 
-        _prepare_location(runpath, location)
-        await _save_record_to_file(record, runpath / location, mime)
+        _prepare_location(root_path, self.location)
+        await _save_record_to_file(record, root_path / self.location, self.mime)
 
-    async def transform_output(self, mime: str, location: Path) -> Record:
+    async def to_record(self, root_path: Path = Path()) -> Record:
```
Comment (jondequinor) [inline ert/data/record/_transformation.py:195]: root_path needs to be None, see above comment

File: ert/data/record/_transformation.py [hunk 4]
```diff
@@ -228,64 +316,82 @@ async def _save_record_to_file(record: Record, location: Path, mime: str) -> Non
             await ft.write(get_serializer(mime).encode(record.data))
     else:
         async with aiofiles.open(str(location), mode="wb") as fb:
-            await fb.write(record.data)  # type: ignore
+            assert isinstance(record.data, bytes)  # mypy
+            await fb.write(record.data)
 
 
-class RecordTreeTransformation(RecordTransformation):
+class RecordTreeTransformation(SerializationTransformation):
```
Comment (jondequinor) [inline ert/data/record/_transformation.py:323]: double check the name, whether or not it make sense

File: ert3/config/_config_plugin_registry.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+class _RegisteredConfig(NamedTuple):
+    config: Type[BaseModel]
+    factory: Callable[[Type[BaseModel]], Any]
+
+
+class ConfigPluginRegistry:
+    def __init__(self) -> None:
+        self._descriminator: Dict[str, str] = {}
+        self._registry: Dict[str, Dict[str, _RegisteredConfig]] = {}
+        self._default: DefaultDict[str, Any] = defaultdict(lambda: Ellipsis)
```
Comment (jondequinor) [inline ert3/config/_config_plugin_registry.py:28]: explain this

File: ert3/config/_config_plugin_registry.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    config_instance = getattr(self, category)
+    if optional and not config_instance:
+        return None
+    elif not optional and not config_instance:
+        raise ValueError("no config, but was required for '{category}' configuration")
+    descriminator_value = getattr(
+        config_instance, plugin_registry.get_descriminator(category=category)
+    )
+    return plugin_registry.get_factory(category=category, name=descriminator_value)(
+        config_instance, parent_config=self
```
Comment (jondequinor) [inline ert3/config/_config_plugin_registry.py:120]: might stamp this `parent_config` out by replacing it with logic that passes `direction` to the transformations configs

File: ert3/config/_ensemble_config.py [hunk 7]
```diff
@@ -34,22 +35,55 @@ class ForwardModel(_EnsembleConfig):
...
     record: str
-    mime: str = ""
-    is_directory: Optional[bool]
-    smry_keys: Optional[List[str]]
+    source: str
 
     @no_type_check
     def __init__(self, **data: Any) -> None:
         super().__init__(**data)
+        if "source" not in data:
```
Comment (jondequinor) [inline ert3/config/_ensemble_config.py:58]: this is old, needs to be replaced by check on "transformation"

File: ert3/config/_experiment_run_config.py [hunk 8]
```diff
@@ -132,34 +129,31 @@ def get_linked_inputs(self) -> Dict[SourceNS, Dict[str, LinkedInput]]:
         stage = self.get_stage()
         for ensemble_input in self._ensemble_config.input:
             name = ensemble_input.record
-            stage_is_directory = stage.input[name].is_directory
-            stage_mime = stage.input[name].mime
-            stage_location = stage.input[name].location
-            stage_smry_keys = stage.input[name].smry_keys
-
-            if stage_mime != ensemble_input.mime:
```
Comment (jondequinor) [inline ert3/config/_experiment_run_config.py:140]: replace with check on affinity

File: ert3/config/_ensemble_config.py [hunk 9]
```diff
@@ -34,22 +35,55 @@ class ForwardModel(_EnsembleConfig):
     driver: Literal["local", "pbs"] = "local"
 
 
+def _ensure_transformation_for_resources(
```
Comment (jondequinor) [inline ert3/config/_ensemble_config.py:38]: needs to be constraint on transformation such that it only is defined for resources

File: ert3/config/_experiment_run_config.py [hunk 10]
```diff
@@ -132,34 +129,31 @@ def get_linked_inputs(self) -> Dict[SourceNS, Dict[str, LinkedInput]]:
...
+                stage.input[name].get_transformation_instance()
+                if stage.input[name].transformation
+                else None
+            )
+            source_transformation: Optional[ert.data.RecordTransformation] = (
+                ensemble_input.get_transformation_instance()
+                if ensemble_input.transformation
+                else None
             )
+            if (
```
Comment (jondequinor) [inline ert3/config/_experiment_run_config.py:143]: double validation

File: ert3/config/_stages_config.py [hunk 11]
```diff
@@ -80,12 +162,13 @@ class _Step(_StagesConfig):
...
 
-    _set_input = validator("input", pre=True, always=True, allow_reuse=True)(
-        _create_record_mapping
-    )
-    _set_output = validator("output", pre=True, always=True, allow_reuse=True)(
-        _create_record_mapping
-    )
+    @validator("input", pre=True, always=True, allow_reuse=True)
+    def _create_input_mapping(cls, ios: List[Dict[str, str]]):
+        return _create_io_mapping(cls, ios, direction="input")
```
Comment (jondequinor) [inline ert3/config/_stages_config.py:167]: to_record

File: ert3/config/_stages_config.py [hunk 12]
```diff
@@ -80,12 +162,13 @@ class _Step(_StagesConfig):
...
+
+    @validator("output", pre=True, always=True, allow_reuse=True)
+    def _create_output_mapping(cls, ios: List[Dict[str, str]]):
+        return _create_io_mapping(cls, ios, direction="output")
```
Comment (jondequinor) [inline ert3/config/_stages_config.py:171]: from_record

File: ert3/console/_console.py [hunk 13]
```diff
@@ -76,12 +76,22 @@ def valid_record_file(path: Union[str, Path]) -> Path:
         help="Indicate that the record is a blob created as a tar from a directory",
     )
 
+    # for python3.8, this should be deprecated and replaced with a --smry-keys
```
Comment (jondequinor) [inline ert3/console/_console.py:79]: out of scope

File: ert3/evaluator/_builder.py [hunk 14]
```diff
@@ -25,13 +26,10 @@ def add_step_inputs(
...
     for input_ in inputs:
-        transformation = _get_input_recordtransformation(input_)
         step_input = (
-            create_file_io_builder()
+            create_input_builder()
             .set_name(input_.name)
-            .set_mime(input_.dest_mime)
-            .set_path(pathlib.Path(input_.dest_location))
-            .set_transformation(transformation)
+            .set_transformation(input_.stage_transformation)
```
Comment (jondequinor) [inline ert3/evaluator/_builder.py:32]: do a check on a direction here

File: ert3/evaluator/_builder.py [hunk 15]
```diff
@@ -77,22 +86,11 @@ async def transform_output(
...
-                mime="application/octet-stream",
                 location=command.location,
             )
         )
-        step.add_input(
-            create_file_io_builder()
-            .set_name(command.name)
-            .set_path(command_location(command.name))
-            .set_mime("application/octet-stream")
-            .set_transformation(ert.data.ExecutableRecordTransformation())
```
Comment (jondequinor) [inline ert3/evaluator/_builder.py:90]: change this back

File: ert3/plugins/hook_specifications/__init__.py [hunk 16]
```diff
@@ -0,0 +1 @@
+from .config import *
```
Comment (jondequinor) [inline ert3/plugins/hook_specifications/__init__.py:1]: enumerate imports

File: ert3/plugins/plugin_manager.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    def __init__(self, plugins=None):
+        super().__init__(_PLUGIN_NAMESPACE)
+        self.add_hookspecs(ert3.plugins.hook_specifications)
+        if plugins is None:
+            self.register(ert3.config.plugins.implementations)
+            self.load_setuptools_entrypoints(_PLUGIN_NAMESPACE)
+        else:
+            for plugin in plugins:
+                self.register(plugin)
+        logging.debug(str(self))
```
Comment (jondequinor) [inline ert3/plugins/plugin_manager.py:27]: rm

File: ert3/plugins/plugin_manager.py [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+
+    def __str__(self):
+        self_str = "ERT Plugin manager:\n"
+        for plugin in self.get_plugins():
+            self_str += "\t" + self.get_name(plugin) + "\n"
+            for hook_caller in self.get_hookcallers(plugin):
+                self_str += "\t\t" + str(hook_caller) + "\n"
+        return self_str
+
```
Comment (jondequinor) [inline ert3/plugins/plugin_manager.py:36]: rm

File: ert3_examples/snake_oil/stages.yml [hunk 19]
```diff
@@ -50,4 +78,4 @@
   script:
     - render -t seed_tmpl -i seed.json -o seed.txt
     - render -t snake_oil_params.txt.tmpl -i *.json -o snake_oil_params.txt
-    - snaky_simulator
+    - snaky_simulator
```
Comment (jondequinor) [inline ert3_examples/snake_oil/stages.yml:81]: ??

File: ert3_examples/snake_oil/stages.yml [hunk 20]
```diff
@@ -2,42 +2,70 @@
...
+      transformation:
+        type: serialization
+        location: seed_tmpl
     -
-      record: snake_oil_params_txt_tmpl
-      location: snake_oil_params.txt.tmpl
+      name: snake_oil_params_txt_tmpl
+      transformation:
+        type: serialization
+        location: snake_oil_params.txt.tmpl
```
Comment (jondequinor) [inline ert3_examples/snake_oil/stages.yml:60]: force numerical here

File: ert_shared/ensemble_evaluator/ensemble/builder.py [hunk 21]
```diff
@@ -153,48 +151,36 @@ def build(self):
...
-        self,
-        name: str,
-        path: Path,
-        mime: str,
-        transformation: Optional[RecordTransformation] = None,
-    ) -> None:
-        super().__init__(name, transformation=transformation)
-        self._path = path
-        self._mime = mime
+# class _FileIO(_IO):
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/ensemble/builder.py:154]: gone

File: ert_shared/ensemble_evaluator/entity/unix_step.py [hunk 22]
```diff
@@ -122,10 +119,10 @@ async def transform_output(output):
 
             futures = []
             for output in self._step.get_outputs():
-                if not (run_path / output.get_path()).exists():
-                    raise FileNotFoundError(
-                        f"Output file {output.get_path()} was not generated!"
-                    )
+                # if not (run_path / output.get_path()).exists():
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/entity/unix_step.py:122]: check this

File: setup.py [hunk 23]
```diff
@@ -91,7 +91,7 @@ def package_files(directory):
         "pluggy",
         "prefect",
         "psutil",
-        "pydantic >= 1.8.1",
+        "pydantic >= 1.9",
```
Comment (jondequinor) [inline setup.py:94]: also in types-requirements?

File: tests/ert_tests/ert3/config/test_config_plugin_registry.py [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+def test_dummy_plugin():
+    plugin_registry = ert3.config.ConfigPluginRegistry()
+    plugin_registry.register_category(category="test", descriminator="type")
+    plugin_manager = ert3.plugins.ErtPluginManager(plugins=[dummy_plugins])
+    plugin_manager.collect(registry=plugin_registry)
+
+    assert plugin_registry.get_factory("test", "foo")
+    assert (
+        str(plugin_registry.get_type("test"))
+        == "<class 'ert3.config._config_plugin_registry.Full_FooConfig'>"
```
Comment (jondequinor) [inline tests/ert_tests/ert3/config/test_config_plugin_registry.py:36]: might import this

File: tests/ert_tests/ert3/conftest.py [hunk 25]
```diff
@@ -378,3 +454,14 @@ def _coerce_raw_ensrec(spec) -> typing.Tuple[ert.data.Record]:
         return tuple(recs)
 
     return _coerce_raw_ensrec
+
+
+@pytest.fixture()
+def plugin_registry():
+    plugin_registry = ert3.config.ConfigPluginRegistry()
```
Comment (jondequinor) [inline tests/ert_tests/ert3/conftest.py:461]: specify internals only

File: ert/data/record/_record.py (hunk 1 above)
Comment (jondequinor) [reply ert/data/record/_record.py:476]: It's immutable, so we've concluded that it's fine as it is.

File: ert3/config/_ensemble_config.py (hunk 7 above)
Comment (jondequinor) [reply ert3/config/_ensemble_config.py:58]: actually, check on `type(self) is Input` which means nothing was plugged in.

File: ert_shared/ensemble_evaluator/entity/unix_step.py (hunk 22 above)
Comment (jondequinor) [reply ert_shared/ensemble_evaluator/entity/unix_step.py:122]: needed to be replaced with `transformation.location` basically

File: setup.py (hunk 23 above)
Comment (jondequinor) [reply setup.py:94]: no?

File: ert/data/record/_transformation.py (hunk 4 above)
Comment (jondequinor) [reply ert/data/record/_transformation.py:327]: @sondreso I don't have a good suggestion for name here… will you help me out?

Comment (sondreso) [reply ert/data/record/_transformation.py:327]: Yhea, thats not an easy one.. TreeSerializationTransformation maybe? :shrug:

**Comment by sondreso** [comment]
Mostly out of curiosity, what does this page (https://ert.readthedocs.io/en/latest/api_reference/config.html) of the documentation look like after this PR?

**Comment by jondequinor** [comment]
> Mostly out of curiosity, what does this page (https://ert.readthedocs.io/en/latest/api_reference/config.html) of the documentation look like after this PR?

It lacks any dynamically produced code, unfortunately.

**Comment by jondequinor** [comment]
test required please!

**Comment by jondequinor** [comment]
test required please!

File: ert3_examples/snake_oil/stages.yml (hunk 20 above)
Comment (jondequinor) [reply ert3_examples/snake_oil/stages.yml:60]: done, failed as it should

File: ert3/plugins/hook_specifications/config.py [hunk 26]
```diff
@@ -0,0 +1,10 @@
+from ert3.plugins.plugin_manager import hook_specification
+from ert3.config import ConfigPluginRegistry
+
+
+@hook_specification  # type: ignore
+def configs(registry: ConfigPluginRegistry) -> None:
+    """
+    This hook allows the user to register plugin configs with the config registry.
+    A Plugin Config must add the class inheriting from ErtScript and an optional name.
```
Comment (sondreso) [inline ert3/plugins/hook_specifications/config.py:9]: Maybe not inherit from `ErtScript`?

**Review by sondreso — COMMENTED** [review]
This is a huge improvement to the code base! Very nice work :clap: When the remaining tests are green I am ready to approve this!

**Review by sondreso — APPROVED** [review]

**Comment by jondequinor** [comment]
Got a comment from Julius that I will need to address, and rebasing on the local run branch revealed a bug. Will address

**Comment by jondequinor** [comment]
ERT - pytest failed due to old pydantic.

**Pull request merged**